Iox 1032 new error handling api

MatthiasKillat opened the pull request:
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

The many files that sometimes only contain little code are intentional  for structure and generic configuration.

At its core the design intention is to keep the types of errors and the backend (i.e. the reaction on errors) very generic without causing much overhead.

Furthermore it shall capture the spirit of the *safe code

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1032

MatthiasKillat commented:
@elBoberido @FerdinandSpitzschnueffler It is still a draft as the following are needed.

1. A usage document (explaining the API use cases)
2. Tests for some internal API (and maybe more tests for the external API
3. More documentation (doxygen?)
4. A design document
5. At least some minor API adjustments to allow tracking the module of an error up to the user-defined code

Conceptually it may still make sense to have a first look, has the overall design is not supposed to change unless there are major issues.

Usage in hoofs will follow in another pull request. It is also an option to add the usage document there. In any case, our error_handling markdown documents requires updating as well.

MatthiasKillat commented:
@elBoberido We also need to update the fatal failure tests once his gets merged. This may be a bit tricky while posh still uses the previous error handling.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,120 @@
+#define IOX_PRECONDITION(expr, message)                                                                                \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        discard([&]() { return expr; });                                                                               \
```
Note that this is a trick to
1. avoid evaluating the expression unless it is needed
2. avoid unused variable warnings in case the whole check is optimized out (`IOX_DEBUG` not set).

We need to consider to that `IOX_DEBUG` as default. Or call switch it around and deactivate the checks with `IOX_UNSAFE`. This is a minor issue and I am fine with both ways.

MatthiasKillat replied:
Not required anymore, removed.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#define IOX_PANIC(...)                                                                                                 \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::panic(CURRENT_SOURCE_LOCATION);                                                                      \
```
pass var args

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/test_hoofs_modules.cpp:
```diff
@@ -32,6 +35,12 @@ int main(int argc, char* argv[])
+    iox::err::TestErrorHandler testHandler;
+    iox::err::ErrorHandler::set(testHandler);
+
+    auto& h = iox::err::ErrorHandler::get();
```
TODO remove

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#define IOX_PANIC(...)                                                                                                 \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::panic(CURRENT_SOURCE_LOCATION);                                                                      \
```
Here `panic` from the platform is directly called, right? If so, one could introduce a `panic` function which doesn't abort execution. Shouldn't we therefore call `forwardError` with `FATAL`?

MatthiasKillat replied:
This is basically the question whether we want to expose `IOX_PANIC` to the user.

`IOX_PANIC` is lower level, it does not need a reason/error on its own to forward, `forwardError` with `FATAL` does.

MatthiasKillat replied:
Good point, must still fix it.

MatthiasKillat replied:
Fixed. Now error_forwarding.hpp takes care of all of the noreturn semantics. (unless something is missing)

MatthiasKillat commented:
@FerdinandSpitzschnueffler @elBoberido I will document more of the design intention today. I also played around with a preliminary integration and found some minor issues with dependencies. I will resolve both of these matters today.

It is probably better to not review until this is done, due to changes and as it hopefully will help understand the structure and configuration capabilities.

elBoberido commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/modules/hoofs/errors.hpp:
```diff
@@ -0,0 +1,71 @@
+using error_code_t = iox::err::error_code_t;
+using module_id_t = iox::err::module_id_t;
+
+enum class ErrorCode : error_code_t
```
With the suggestion from above we could have a `error_code_t::type` ... maybe we should drop the `_t` when `error_code_t` is a distinct type.

elBoberido replied:
You got me wrong. I did not suggest to replace the enum but with the suggestion to make `error_code_t` a distinct type instead of an alias the enum would not work anymore. The workaround would be to add a typedef to the `error_code_t` struct.

elBoberido commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/modules/hoofs/errors.hpp:
```diff
@@ -0,0 +1,71 @@
+namespace err
+{
+
+inline hoofs_errors::Error toError(hoofs_errors::ErrorCode code)
```
Shouldn't this be `error_code_t`?

elBoberido reviewed (commented):
Reviewed the first 10 file. I think we have to abstain the [primitive obsession](https://refactoring.guru/smells/primitive-obsession)

MatthiasKillat commented on doc/design/error_reporting_api.md:
```diff
@@ -0,0 +1,240 @@
+A unique id for each module that is defined by some central instance that knows all modules that are
+to be used in some project.
+
+### Platform 
+
+The backend code that is used by any module. It can be changed at compile time and there is only
+one active platform at a time.
```
Yes the name is from a time where the platform abstraction was still discussed. It is the compile time configuration part.

MatthiasKillat replied:
Renamed to custom as it is the custom implementation part.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error.hpp:
```diff
@@ -0,0 +1,103 @@
+}
+
+template <class Error>
+inline error_code_t toModule(const Error& error)
```
Can do. Effort depends on whether we need to test all these small types though.

MatthiasKillat replied:
Yes, should create custom types for that instead of alias to primitive types.

MatthiasKillat replied:
Abstracted in `types.hpp`. No fundamental types should be exposed in the user or custom API any more.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#ifndef IOX_HOOFS_ERROR_REPORTING_API_HPP
```
Will do.

MatthiasKillat replied:
Done.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_kind.hpp:
```diff
@@ -0,0 +1,130 @@
+template <typename Kind>
+bool constexpr requiresHandling(Kind)
+{
+    return true;
```
No, the intention is that the primary template assumes it has to be handled, unless explicitly specified otherwise. It is a opt in vs. opt out question.

MatthiasKillat replied:
Removed this trait entirely.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/modules/hoofs/errors.hpp:
```diff
@@ -0,0 +1,71 @@
+  protected:
+    ErrorCode m_code;
+
+    static constexpr const char* errorNames[] = {"Unknown"};
```
Probably, but it must be specified somewhere (and I suggest some `errors.hpp` file that is specific to the module).

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_logging.hpp:
```diff
@@ -0,0 +1,24 @@
+
+/// @brief Log the location of an error.
+/// @param location the location of the error
+#define IOX_LOG_ERROR(location)                                                                                        \
```
Sure. Actually we could do without macros here later if the logging API is sufficiently exposed (we already got location).

MatthiasKillat commented on doc/design/error_reporting_api.md:
````diff
@@ -0,0 +1,240 @@
+```cpp
+int x;
+// ...
+IOX_ASSERT(x>=0, Code::OutOfBounds)
````
Sure. The point is that an assumption is not necessarily checked, an assertion is (here...). `IOX_REQUIRE` should make this more clear.

MatthiasKillat replied:
Renamed. But we it has the disadvantage that it is a new "keyword". Not important to me though, I am concerned with the purpose mainly.

MatthiasKillat commented on doc/design/error_reporting_api.md:
```diff
@@ -0,0 +1,240 @@
+
+The frontend to repor errors. API for short.
+
+### Error Handler
+
+The backend that defines the actions on reported errors. Defined by the platform.
+
+### Error Code
+
+Some error code that is unique per module.
+
+### Error
+
+An error object to be reported. Must contain at least some error code and is associated with some
+module (by its id).
```
Sure. Depends on how general this should be but we should agree on the scope first.

MatthiasKillat replied:
I did not incorporate this yet. Would prefer to do this in a follow up. It would be mostly renaming I think.

MatthiasKillat commented on doc/design/error_reporting_api.md:
```diff
@@ -0,0 +1,240 @@
+All of them must be supported by the platform and all of them abort execution (which cannot be
+changed).
+
+Furthermore the platform may define additional kinds of errors, specifically non-fatal errors.
+For non-fatal errors the execution continues afer reporting, so they should be used when the
+circumstances allow recovery from the error.
```
This is intended for what we previously had in posh as warnings etc.

My idea is to support three steps.

1. Generate the error `e` somehow (`IOX_ERROR`, code, get it by return value from somewhere etc.)
2. Report, but maybe not fatal.
3. If not fatal, control flow continues (likely differently) and `e` is returned or thrown or whatever.

Maybe there is a misunderstanding. The platform can add custom kinds, the user is not supposed to do this. They have to use what the platform provides, otherwise the code will not compile. This is similar to trying to throw a `std::runtime_error` but it is not defined. We rely on this being supported by the "platform", the C++ lib in this case.

elBoberido replied:
It is not only the user which has to use whatever the platform provides but every platform hast to provide whatever the user uses. Not specifying this repeats one of the big mistakes of C++ by leaving stuff unspecified and let the vendors decide. Neither the user nor the provider of the platform has full freedom and by not specifying it we make their life harder

elBoberido commented on iceoryx_hoofs/test/test.hpp:
```diff
@@ -21,4 +21,9 @@
 #include <gtest/gtest.h>
 #include <memory>
 
+// defensive checks (IOX_ASSUME, IOX_PRECONDITION) are active during testing
+/// @todo this is preferably configured in cmake
+#define IOX_CHECK_PRECONDITIONS
+#define IOX_CHECK_ASSUMPTIONS
+
```
Please remove this from here. It is easily forgotten when we add a cmake configuration. Just use what is defined in `hoofs/error_reporting.hpp` else we might run into trouble in the future.

I would also move away from the `define` and create a `constexpr bool IOX_CHECK_PRECONDITIONS {true};` then it can more easily be configured via `iceoryx_hoofs/cmake/iceoryx_hoofs_deployment.hpp.in`

MatthiasKillat replied:
Configured in `configuration.hpp` now which contains a bunch (two ...) compile time constants and can be extended as needed.

elBoberido reviewed (commented):
Reviewed everything but `test_error_handling_api.cpp`

MatthiasKillat commented on iceoryx_hoofs/CMakeLists.txt:
```diff
@@ -105,6 +105,8 @@ iox_add_library(
         source/posix_wrapper/unix_domain_socket.cpp
         source/memory/relative_pointer_data.cpp
         time/source/duration.cpp
+        source/error_reporting/platform/default/default_error_handler.cpp
+        source/error_reporting/platform/default/test_error_handler.cpp
```
I restructure tomorrow. Lots of good suggestions, but limited time.

I still think at its core this design is the way to go. We can adapt stuff to report at compile time, runtime, configure reaction. It just looks not super intuitive to understand, but the user does not have to. Only the implementer of some platform etc.

MatthiasKillat replied:
Moved.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/error_reporting.hpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+[[noreturn]] inline void panic(const SourceLocation& location, const char* msg)
+// inline void panic(const SourceLocation& location, const char* msg)
```
Apparently I am in a rush to get this done :-(

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/error_reporting.hpp:
```diff
@@ -0,0 +1,81 @@
+    abort();
+}
+
+template <class Kind, class Error>
+inline void report(const SourceLocation& location, Kind, const Error& error)
+{
+    auto code = toCode(error);
```
We probably can, but I wanted to prepare everything for custom errors (per platform, not module!).

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/error_reporting.hpp:
```diff
@@ -0,0 +1,81 @@
+{
+    auto code = toCode(error);
+    IOX_LOG_FATAL_ERROR(location) << " Fatal Error " << code.value << " in module " << toModule(error).value;
+    auto& h = ErrorHandler::get();
```
Make this a one liner; always.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/test_error_handler.hpp:
```diff
@@ -0,0 +1,76 @@
+    std::atomic<bool> m_panicked{false};
+    std::vector<error_code_t> m_errors;
+
+    jmp_buf m_jumpBuffer;
+    bool m_jump{false};
```
But the `testhandler` must execute the jump in panic. It needs at least some knowledge about this.

MatthiasKillat replied:
It is now secured in a way that it should be multi-threading enabled in the sense that only one thread can obtain the buffer. The jump must still be to a valid stackframe in the thread that obtained the buffer, we cannot enforce this due to API limitations of `setjmp`. Note that this is in internal code (the whole `TestErrorHandler`) and we provide further abstractions in `test_support.hpp` (that should be used in test code).

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/test_error_handler.hpp:
```diff
@@ -0,0 +1,76 @@
+    bool hasError() const;
+
+    /// @todo module id must be propagated here as well
+    /// @brief Indicates whether as specific error occurred previously.
```
```suggestion
    /// @brief Indicates whether a specific error occurred previously.
```

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/minimal/error_reporting.hpp:
```diff
@@ -0,0 +1,63 @@
+#ifndef IOX_HOOFS_ERROR_REPORTING_PLATFORM_MINIMAL_ERROR_REPORTING_HPP
```
The intention is to showcase the minimum API that must be implemented to work/do something useful.

MatthiasKillat replied:
Removed.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_kind.hpp:
```diff
@@ -0,0 +1,130 @@
+    return true;
+}
+
+template <typename Kind>
+bool constexpr requiresHandling(Kind)
```
Yes. It allows additional configuration at the platform level but must be used carefully (not possible for any fatal error, it is predefined for these).

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#define IOX_ASSERT(expr, error) IOX_REPORT_IF(!(expr), error, FATAL)
+
+//*****************************
+//* For safe mode and debugging
```
Sure. You know my vision of adding more contract specific stuff to the project and most importantly properly distinguish between `bug conditions` (i.e. not supposed to happen) and `error conditions` (can happen and must be handled).

Not saying it is perfect by any means, but it is the journey that counts. :-)

MatthiasKillat replied:
Action needed?

elBoberido replied:
I'm just wondering if `safe` is the right word here. I mean why would someone not use the `safe mode`. It is an `all checks enabled mode` but I don't have a catchy name for it

MatthiasKillat commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_support.hpp:
```diff
@@ -0,0 +1,133 @@
+    if (t.joinable())
+    {
+        t.join();
+    }
```
else there should be a test failure, this should not happen

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+///        default specification or override by module
+/// @code error code to be transformed
+/// @note this relies on overloading
+#define IOX_ERROR(code) iox::err::toError(code)
```
Removed.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#define IOX_PANIC(...)                                                                                                 \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::panic(CURRENT_SOURCE_LOCATION);                                                                      \
```
Done, was renamed `source_location.hpp` though for clarity.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/error_kind.hpp:
```diff
@@ -0,0 +1,38 @@
+    ERROR = FATAL_LEVEL + 1,
+};
+
+// prefer types to avoid switch statements and the like and allow annotations (such as name here)
```
Removed. It was about that I prefer to use the type system (and overloading/generics) to switch case for multiple reasons. It is IMO the clean way more extendable.

MatthiasKillat commented on doc/design/error_reporting_api.md:
```diff
@@ -0,0 +1,240 @@
+There are three mandatory kinds (or categories) of errors.
+
+1. Fatal errors
+1. Precondition violations
```
No. This would be an assumption violation/debug assert violation. Clarified?

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+            {                                                                                                          \
+                forwardError(CURRENT_SOURCE_LOCATION, IOX_ERROR(error), kind);                                         \
+            }                                                                                                          \
+        }                                                                                                              \
```
Combined. But I would think it is optimized fairly heavily. Not that these metric tools care about this ...

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,125 @@
+#define IOX_REPORT_IF(expr, error, kind)                                                                               \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (requiresHandling(iox::err::kind))                                                                          \
```
Not sure what you mean but there were extensive changes so maybe it is resolved.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/error_handler_interface.hpp:
```diff
@@ -0,0 +1,34 @@
+    /// @note some of the code that is always supposed to be exectuted is factored out at call-site
+    /// @note we distinguish the error by error_code_t, alternatively we would need an error
+    /// class hierarchy which would be heavier and is not needed now (can be extended in the future)
+    virtual void report(const SourceLocation& location, error_code_t code) = 0;
```
Yes. TBD.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_forwarding.hpp:
```diff
@@ -0,0 +1,120 @@
+        {
+            forwardNonFatalError(location, std::forward<Error>(error), std::forward<Kind>(kind));
+        }
+    }
```
Changed.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/default/error_reporting.hpp:
```diff
@@ -0,0 +1,81 @@
+    IOX_LOG_FATAL_ERROR(location) << ": Debug Assert Violation " << std::forward<Message>(msg);
+    auto& h = ErrorHandler::get();
+    h.report(location, code);
+}
```
TBD.

MatthiasKillat replied:
Not done yet. I am not sure yet how this ideally would look like. Besides I want to avoid `function_ref` dependency.

MatthiasKillat replied:
Not really an option, as it cannot depend on `function_ref`. The custom implementation can be improved/generalized in a follow-up. For now I consider it good enough.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/error_kind.hpp:
```diff
@@ -0,0 +1,38 @@
+};
+
+// prefer types to avoid switch statements and the like and allow annotations (such as name here)
+struct Error
```
Changed much here, should not apply anymore.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/platform/error_reporting.hpp:
```diff
@@ -0,0 +1,30 @@
+// 2. Can be exchanged at runtime
+// 3. Logs errors
+
+#include "iceoryx_hoofs/error_reporting/platform/default/error_reporting.hpp"
```
Removed the switch entirely. Always include the default for now. We can think about the mechanism to exchange it at compile time later. It boils down to select the right header, ideally just one.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/module_a/errors.hpp:
```diff
@@ -0,0 +1,100 @@
+inline module_a::errors::Error toError(module_a::errors::ErrorCode code)
+{
+    return module_a::errors::Error(code);
+}
```
The default implementation will returns the same object (maybe there should be no default for safety?). This works, as long as the type returned implements `code` and `module` (and maybe later some other functions).

This is to make it work out of the box for any type conforming to the interface.

This is not the case for an enum like here, hence it would break at compile time. This specialization takes care of it by defining how to construct an error from a code. I see no issue here, except that the implicit default identity behavior may be undesirable (a matter of preference to me).

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_helper.hpp:
```diff
@@ -0,0 +1,77 @@
+#include <functional>
+#include <stdio.h>
+
+#define ASSERT_NO_PANIC()                                                                                              \
```
It is supposed to be generic for all tests but maybe not fully mature. We need more experience with usage and hence there may be changes. Moved it as requested and renamed it `test_support.hpp`

MatthiasKillat replied:
@Reworked. See `test_support.hpp`.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_helper.hpp:
```diff
@@ -0,0 +1,77 @@
+    {
+        testFunction(std::forward<Args>(args)...);
+    }
+}
```
Yes. We talked about this during some coffee break. But this cannot be fixed IMO due to limitations of `setjmp`. `setjmp` and `longjmp` must happen in the same thread. I see no way to enforce this except stating limitations.

elBoberido replied:
It is a long time since I reviewed this but the easiest solution is to not have the jmp buf as part of the error handler but just as local variable. Some things would require a refactoring though.

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+This has advantages for test coverage as the compiler and other tools that rely on the compiler (say
+for coverage) are aware of the `noreturn` guarantee of `IOX_UNREACHABLE`.
+As a consequnece, branches with `IOX_UNREACHABLE` do not necessarily lead to a return statement.
```
```suggestion
As a consequence, branches with `IOX_UNREACHABLE` do not necessarily lead to a return statement.
```

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+int identity()
+{
+    auto result = algorithm();
```
Same here.

MatthiasKillat replied:
Same as above. Will change.

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+### Basics
+
+The overall implementation concepts allows customizationof the implementation level and a provides 
```
```suggestion
The overall implementation concepts allow customization of the implementation level and provide
```

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+The default implementation in `custom/default` allows switching between a `DefaultHandler` 
+and a `TestHandler` at runtime.
+The latter is used in testing to verify that an error occured when the it is expected.
```
```suggestion
The latter is used in testing to verify that an error occurred when it is expected.
```

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+These are the definition of `TestHandler` in `test_error_handler.hpp` and auxiliary functions in
+`test_support.hpp` to be used in tests to verify errors. The latter can be extended as required.
+
+### Custom Implementation
+
+Extension of existing definitions is possible by either changing the default implementation 
+or providing an additional custom implementation and ensure that it is used by all modules.
```
This should be combined with the `Custom Implementation` section above.

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+1. Adapting the error definitions for `iceoryx_hoofs` in `modules/hoofs/errors.hpp`
+2. Introducing a similar folder structure for `iceoryx_posh`
+3. Replacing occurences of the previous error handler call (including `cxx::Expects`)
```
```suggestion
3. Replacing occurrences of the previous error handler call (including `cxx::Expects`)
```

FerdinandSpitzschnueffler reviewed (commented):
Only reviewed the readme so far.

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+    {                                                                                                                  \
+        iox::err::forwardPanic(CURRENT_SOURCE_LOCATION, message);                                                      \
+    } while (false)
+/// @brief report error of some kind
```
Please add a blank line above.

FerdinandSpitzschnueffler commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+### Assert That a Condition Holds
+
+Similarly we can conditionally check whether a condition does not hold and report a fatal error in
+the case that it does not
```
```suggestion
Similarly we can conditionally check whether a condition does hold and report a fatal error in
the case that it does not
```
Otherwise it is like a double negation.

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+        }                                                                                                              \
+    } while (false)
+
+/// @brief if enabled: report (fatal) assumption violation if expr evaluates to false
```
Why is fatal in parenthesis? There is always a fatal error reported.

MatthiasKillat replied:
True. It was more that I thought it is not of immediate interest (extra info). Will remove parens.

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+                                        message);                                                                      \
+        }                                                                                                              \
+    } while (false)
+/// @brief panic if control flow reaches this code at runtime
```
Please add a blank line above.

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/custom/error_kind.hpp:
```diff
@@ -0,0 +1,43 @@
+namespace err
+{
+
+// The non-fatal errors kinds can all be defined here.
```
```suggestion
// The non-fatal error kinds can all be defined here.
```

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/errors.hpp:
```diff
@@ -0,0 +1,114 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_ERROR_REPORTING_ERROR_HPP
```
```suggestion
#ifndef IOX_HOOFS_ERROR_REPORTING_ERRORS_HPP
```

FerdinandSpitzschnueffler reviewed (commented):
Reviewed everything but the tests.

elBoberido reviewed (commented):
Finished with the design document

elBoberido commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/modules/hoofs/errors.hpp:
What do you think of having a list of all known module IDs? This would make it easier for us to keep track of all module IDs and prevent to reuse an ID. The downside is of course to mention something like `posh` or `dust` in `hoofs` but I think the benefits outweigh the little ugliness. It would be similar to IANA having a curated list of ports.

MatthiasKillat replied:
Will do in an integration follow up.

elBoberido commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/errors.hpp:
```diff
@@ -0,0 +1,114 @@
+    return error.code();
+}
+
+template <>
+inline ErrorCode toCode<ErrorCode>(const ErrorCode& error)
+{
+    return error;
+}
```
I guess the purpose of this specialization is template meta-programming?

MatthiasKillat replied:
It is needed as every error type (and the ErrorCode is the most simple one) must define a conversion to the ErrorCode (by overload/specialization). Otherwise the code does not compile.

MatthiasKillat replied:
In a way this is all "concept" based but without using concepts, because we cannot for multiple reasons: C++ standard, here we actually prefer to have a free function to allow primitive types (such as `ErrorCode` as a thin wrapper) etc.

elBoberido reviewed (commented):
`tests`/`testing` is not yet reviewed

MatthiasKillat commented on .clang-tidy-diff-scans.txt:
```diff
@@ -25,6 +25,11 @@
+./iceoryx_hoofs/include/error_reporting/**/*
+./iceoryx_hoofs/source/error_reporting/**/*
+./iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_error_handler.hpp
+./iceoryx_hoofs/testing/error_reporting/test_error_handler.cpp
```
The original thought was that some of the test code does not necessarily have to obey the clang tidy rules but IMO at least the handler itself even if part of test code) should be checked.

If extending it to all test files does not lead to too many suppression I will add the folder as well.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+#define IOX_UNREACHABLE()                                                                                              \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::panic(CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.");                 \
```
Right, it must be like you mentioned. Oversight.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+expected<int, E> identity()
+{
+    auto result = algorithm();
```
Why? I mean in theory it could be failing due to side effect but I will add an argument. 

Reminds me of a Monty Python sketch actually. I like to have an argument...

MatthiasKillat replied:
The naming `identity` was off as well, the example changed but the naming did not. Updated.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_forwarding.hpp:
```diff
@@ -0,0 +1,84 @@
+template <typename Error, typename Kind>
+[[noreturn]] inline void forwardFatalError(Error&& error, Kind&& kind, const SourceLocation& location)
+{
+    report(location, std::forward<Kind>(kind), std::forward<Error>(error));
```
Because I was trying out some tricks to keep variadic macro arguments which ultimately failed (not compliant with standard C). Then the different order was beneficial but I forgot why...

Before this it was in the order you mentioned.
 Now it does not matter and and I tend to keep it as is (as you said).

MatthiasKillat commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_error_handler.hpp:
```diff
@@ -0,0 +1,115 @@
+#include "iceoryx_hoofs/error_reporting/source_location.hpp"
+#include "iceoryx_hoofs/error_reporting/types.hpp"
+
+#include <algorithm>
```
No. Removed.

MatthiasKillat commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_support.hpp:
```diff
@@ -0,0 +1,158 @@
+#define EXPECT_PRECONDITION_VIOLATION() EXPECT_TRUE(iox::testing::hasPreconditionViolation())
+
+#define EXPECT_ASSUMPTION_VIOLATION() EXPECT_TRUE(iox::testing::hasAssumptionViolation())
+
```
Matter of opinion I guess. I can remove them but I like them for expressiveness (intentional programming and a little shorter) and they can be more easily searched for.

elBoberido replied:
The problem is that it is hard to remember what is available. Something like
```cpp
using iox::testing;
...
EXPECT_TRUE(ErrorHandler::hasViolation());
```
might be easier to use.

On the other hand, a proper IDE should give some good hints for `IOX_EXPECT_`/`IOX_TESTING_EXPECT_`. For `EXPECT_` only, there are already quite a few gTest macros.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_auxiliary_types.cpp:
```diff
@@ -0,0 +1,132 @@
+    EXPECT_FALSE(this->sut == different);
+}
+
+TYPED_TEST(RegularType_test, unequalityComparisonWorks)
```
https://www.merriam-webster.com/words-at-play/unequal-vs-inequality-negative-prefix-usage

I prefer to change it to `unequalComparisonWorks` then. For me equality is related to =, unequal(ity) to != and inequality to <, >, <=, >=.

Language is inconsistent.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_errors.cpp:
```diff
@@ -0,0 +1,206 @@
+    EXPECT_FALSE(this->sut == different3);
+}
+
+TYPED_TEST(ErrorType_test, unequalityComparisonWorks)
```
See above. Will change.

MatthiasKillat commented on doc/design/error_reporting.md:
````diff
@@ -0,0 +1,456 @@
+}
+```
+
+If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 
````
It is not possible anymore. The documentation is inconsistent and will be changed.

MatthiasKillat replied:
Changed.

MatthiasKillat replied:
@FerdinandSpitzschnueffler It should not anymore, but maybe I missed it while reading the paragraph again.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+interface `error_reporting.hpp`. 
+
+Apart from implementing the error reporting interface, a custom implementation does not have to follow 
+a specific structure. However, it cannot depend on anything that intends to use the error 
+reporting itself. It can override or extend some definitions and it is encouraged to use the same 
```
Yes. This is a problem I also encountered when plying with support for expected.

MatthiasKillat commented on iceoryx_hoofs/design/include/iox/static_lifetime_guard.hpp:
```diff
@@ -78,7 +78,14 @@ class StaticLifetimeGuard
 
   private:
-    using storage_t = typename std::aligned_storage_t<sizeof(T), alignof(T)>;
+    struct alignas(T) storage_t
+    {
+        // AXIVION Next Construct AutosarC++19_03-M0.1.3 : the field is intentionally unused and serves as a mean to provide memory
+        // AXIVION Next Construct AutosarC++19_03-A1.1.1 : object size depends on template parameter and has to be taken care of at the specific template instantiation
+        // AXIVION Next Construct AutosarC++19_03-A18.1.1 : required as low level building block, encapsulated in abstraction and not directly used
+        // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+        unsigned char data[sizeof(T)];
+    };
```
Actually `aligned_storage`. But not here. It is only because of a MSVC compiler bug/legacy fix.

MatthiasKillat replied:
Nothing to do here.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
Fine with me. Note I do not like prefixing `error_reporting` and prefer using the folder structure but just `macros` would be a worse name.

MatthiasKillat replied:
Changed.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+#define IOX_REPORT(error, kind)                                                                                        \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (iox::err::isFatal(kind))                                                                                   \
+        {                                                                                                              \
+            iox::err::forwardFatalError(iox::err::toError(error), kind, CURRENT_SOURCE_LOCATION);                      \
+        }                                                                                                              \
+        else                                                                                                           \
+        {                                                                                                              \
+            iox::err::forwardNonFatalError(iox::err::toError(error), kind, CURRENT_SOURCE_LOCATION);                   \
+        }                                                                                                              \
```
Can do.

MatthiasKillat replied:
I have to think more about this and will elaborate. It is not that simple.

MatthiasKillat replied:
Changed as per our discussion.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+4. `error_kind.hpp`: mandatory error categories
+5. `error_logging.hpp`: logging related definitions
+6. `location.hpp`: source location related definitions
+7. `types.hpp`: auxiliary types
```
Added.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+
+These are
+
+1. `api.h`: the reporting API to be used
```
I will change the name. It was just to make the distinction between (public) API and auxiliary code clear.

MatthiasKillat replied:
Changed as requested.

MatthiasKillat commented on doc/design/error_reporting.md:
````diff
@@ -0,0 +1,456 @@
+```
+
+is used to verify assumptions **BEFORE** any logic in the function body is executed. Technically copy
+constructors may run before any condition can be checked, and there is also the possibility of
+reordering if the following code does not depend on the condition at all. This is a limitation of
+the language and cannot be avoided.
````
Yes but the compiler may move code below the precondition check (without observable effects including side effects) before the precondition check unless I miss something. So it is not a problem, but still not as it should work for "perfect precondition checking", namely just before the function is called.

MatthiasKillat commented on doc/design/error_reporting.md:
````diff
@@ -0,0 +1,456 @@
+}
+```
+
+If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 
````
Done

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_forwarding.hpp:
```diff
@@ -0,0 +1,84 @@
+#include "iceoryx_hoofs/error_reporting/error_kind.hpp"
+#include "iceoryx_hoofs/error_reporting/source_location.hpp"
+
+// to establish connection to the custom implementation
+#include "iceoryx_hoofs/error_reporting/custom/error_reporting.hpp"
```
Yes. This is not the best solution and we can improve this somehow with Cmake magic I think (but did not consider what the best mechanism would be). The crucial point is that the header has to be selected at build definition time or compile time at the latest.

The  non-custom code part knows it has to look into `iceoryx_hoofs/error_reporting/custom/error_reporting.hpp` for the definition (a compile time contract). Failing this, it will not work.

For now it is good enough. Can be improved once we consider more customization.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/custom/default/default_error_handler.hpp:
```diff
@@ -0,0 +1,60 @@
+    void onReportError(ErrorDescriptor desc) override;
+
+    /// @brief Defines the reaction onviolation.
+    /// @param location the location of the violation
```
True. Part of the descriptor because this is easier to change. The interface for the default handler is a customization design decision and not mandatory of course.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/custom/default/default_error_handler.hpp:
```diff
@@ -0,0 +1,60 @@
+    void onPanic() override;
+
+    /// @brief Defines the reaction on error.
+    /// @param location the location of the error
```
It was before. Part of the descriptor now because this is easier to change. The interface for the default handler is a customization design decision and not mandatory of course.

MatthiasKillat commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_support.hpp:
```diff
@@ -0,0 +1,158 @@
+
+// NOLINTBEGIN(cppcoreguidelines-macro-usage) macro required for source location in tests
+
+#define ASSERT_IOX_OK() ASSERT_TRUE(iox::testing::isInNormalState())
```
The intention was to convey that the IOX error state is ok but I am not happy with it either (but neither with `ASSERT_OK`).

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_error_reporting.cpp:
Reasonable. Changed.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_error_reporting_api.cpp:
```diff
@@ -0,0 +1,266 @@
+    EXPECT_PANIC();
+}
+
+TEST_F(ErrorReportingApi_test, reportNonFatal)
```
Added for now. After some discussion with @elBoberido I came to the conclusion that we do not want to allow this to work (due to `IOX_REPORT_FATAL`) and subsequent return statements that must not exist in the fatal case (but have to in non fatal cases). 

I am of the opinion that if it has to be fatal `IOX_REPORT_FATAL` should be used and no subsequent return statement is needed. If only `IOX_REPORT` is used, a subsequent return statement is required regardless of whether the error is fatal or not. 

However, this can break code if an error kind changes from fatal to non-fatal if this is not enforced (which cannot be  done).

I prefer that it would work and hence added the test but I will think about it. I think it is not immediately harmful either way.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/error_reporting/test_error_reporting_api.cpp:
```diff
@@ -0,0 +1,266 @@
+    runInTestThread(f);
+
+    EXPECT_IOX_OK();
+}
```
Added. But similar considerations apply. Maybe we should restrict it but I prefer not too. We need more usage experience IMO.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/custom/default/configuration.hpp:
```diff
@@ -0,0 +1,40 @@
+namespace err
+{
+
+// Specialize to change the checks (and other options if needed) at compile time.
+// this can later also be done depending on a #define to select a header
+// but we should avoid to have a #define for each option.
```
I propose to do this in a follow up. This is a deployment/build configuration issue. I have not considered many options, but it has to be isolated to few (ideally one) file and not end up in a bunch of `#defines` if possible.

MatthiasKillat replied:
Nothing to be done here as of now.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/types.hpp:
```diff
@@ -0,0 +1,82 @@
+
+    type value;
+
+    static constexpr type ANY{0};
```
Added but is not used. Can be put to test in first integration with the base library.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp:
```diff
@@ -0,0 +1,140 @@
+            else                                                                                                       \
+            {                                                                                                          \
+                iox::err::forwardNonFatalError(iox::err::toError(error), kind, CURRENT_SOURCE_LOCATION);               \
+            }                                                                                                          \
```
Changed as per our discussion.

MatthiasKillat commented:
@elBoberido @FerdinandSpitzschnueffler I think I fixed all the critical problems, but please double check. I mentioned were I would defer actions until integration or where no easy change is possible.

Most importantly I added a general way to translate module names and error names, provided the error definition file of the module defines them.

Furthermore I restricted the behaviour of `IOX_REPORT` and `IOX_REPORT_IF` to non-fatal errors only and created `FATAL` counterparts, as discussed. The documentation was updated accordingly.

I think to iron out further issues an actual integration is required. I am confident that it is not too difficult for hoofs and also think that these definitions combined with monadic types (or exceptions) enable an idiomatic expressive coding style suitable for a safety context. Time will tell how this works out.

FerdinandSpitzschnueffler commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/modules/hoofs/error_reporting.hpp:
```diff
@@ -0,0 +1,31 @@
+#include "iceoryx_hoofs/error_reporting/custom/error_reporting.hpp"
+
+// 3. Include the error reporting API
+#include "iceoryx_hoofs/error_reporting/api.hpp"
```
Wasn't this renamed?

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_reporting_macros.hpp:
```diff
@@ -0,0 +1,141 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_ERROR_REPORTING_API_HPP
```
It must be. And that is one downside of include guards... (one can have automation for it, but still ...)

MatthiasKillat replied:
Done

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_reporting_macros.hpp:
```diff
@@ -0,0 +1,141 @@
+#define IOX_HOOFS_ERROR_REPORTING_API_HPP
+
+#include "iceoryx_hoofs/error_reporting/configuration.hpp"
+#include "iceoryx_hoofs/error_reporting/custom/error_kind.hpp"
```
We do not. To be removed.

elBoberido commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+It should not be used at the start of a function body and instead replaced with a precondition check
+in this case.
+
+### Marking Unreachable Code
```
This should not be under a section which states that it can be disabled at compile time, like `## Using the API to Check Contracts and Assumptions` indicates

MatthiasKillat replied:
Put into its own section (reduced header indent level).

elBoberido commented on doc/design/error_reporting.md:
````diff
@@ -0,0 +1,456 @@
+
+It is also possible to explicitly state that code is supposed to be unreachable.
+```cpp
+if(condition) {
+    // Reachable code that does something
+    // This also implies that it is assumed that the condition cannot be false.
+} else {
+    IOX_UNREACHABLE();
+    // Code here should be dead, otherwise it is a bug
+    // There should ideally be no dead code, but there are exceptions.
+}
````
A better example would be an exhausting switch statement with a return in each case and an `IOX_UNREACHABLE` after the switch statement. This is a real world use case since one can cast any number to an enum class.

MatthiasKillat replied:
I added the example but did not remove the previous one because I do not agree it is better, it is just another use case.

elBoberido commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+    {
+        IOX_REPORT(SomeError, RUNTIME_ERROR);
+        // control flow continues and the error is propagated to the caller
+        return error<Code>(SomeError);
```
```suggestion
        return err(SomeError);
```
There is an open PR which changes the API

elBoberido reviewed (commented):
Intermediate re-re-review :sweat_smile:

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+8. `errors.hpp` : supported error types and related free functions
+
+All the files focus on singular aspects to allow fine-grained inclusion.
+All definitions have to reside in `iox::err`, which is considered a private (detail) namespace 
```
I see an renaming of my namespace on the horizon. OK, I think about a name. No big deal. Will be done after a complete review.

MatthiasKillat replied:
Renamed `iox::err` to `iox::er` as discussed offline.

elBoberido commented on iceoryx_hoofs/testing/error_reporting/test_error_handler.cpp:
Just a suggestion. Since all our tests files start with `test_`, what do you think of naming this `testing_`?

MatthiasKillat replied:
OK with me. But no class name change.

elBoberido commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_support.hpp:
```diff
@@ -0,0 +1,158 @@
+
+    if (buf == nullptr)
+    {
+        // should not fail with correct usage
+        GTEST_FAIL();
```
```suggestion
        GTEST_FAIL() << "This should not fail! Incorrect usage!";
```

elBoberido commented on iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/error_reporting/test_support.hpp:
```diff
@@ -0,0 +1,158 @@
+
+// NOLINTBEGIN(cppcoreguidelines-macro-usage) macro required for source location in tests
+
+#define ASSERT_OK() ASSERT_TRUE(iox::testing::isInNormalState())
```
All of these macros need to be prefixed with at least `IOX_`. Even better would be `IOX_TESTING_` to not confuse the user with the `IOX_ASSUME` macro.

MatthiasKillat replied:
Done

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/custom/default/error_reporting.inl:
```diff
@@ -0,0 +1,141 @@
+{
+    auto code = toCode(error);
+    auto module = toModule(error);
+    IOX_ERROR_INTERNAL_LOG(location) << " Error " << code.value << " in module " << module.value;
```
Done.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,456 @@
+        return into<error<E>>(result.get_error());
+    }
+    // no error, return result
+    return *result;
```
But we do not as of now as far as I am aware so I changed it as suggested.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/error_reporting_macros.hpp:
```diff
@@ -0,0 +1,141 @@
+/// @param message message to be forwarded
+/// @note could actually throw if desired without breaking control flow asssumptions
+#define IOX_PANIC(message)                                                                                             \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::forwardPanic(CURRENT_SOURCE_LOCATION, message);                                                      \
+    } while (false)
```
No. We will tackle this once it becomes an issue and use idiomatic macros for now to prevent using them as expressions.

Using cyclomatic complexity with hard thresholds is foolish anyway (and we do not).
Usually it is easy to spot bad functions with high complexity in reviews anyway.

MatthiasKillat commented:
@FerdinandSpitzschnueffler @elBoberido Ready for hopefully final review. Adapted namespaces, markdown file and testing macros and file names.

FerdinandSpitzschnueffler reviewed (commented):
IMO this can be merged once the CI is fixed.

elBoberido reviewed (commented):
Just a tiny last issue in an example

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -249,6 +252,26 @@ if(condition) {
+void handleColor(Color color) {
+
+    switch(color) {
+        case Color:Red { handleRed(); }
+
+        case Color:Blue { handleBlue(); }
```
True, oversight on my part. I should not do dry coding, even if it it is trivial.

MatthiasKillat commented on doc/design/error_reporting.md:
```diff
@@ -0,0 +1,481 @@
+        case Color:Red { return handleRed(); }
+
+        case Color:Blue { return handleBlue(); }
+    }
```
Yes ... I am not hat my best today. Fixed.
